[WIP][dashboard scoped filter] part 1: scope selector modal

graceguo-supercat opened the pull request:
### CATEGORY

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
This is part 1 for Dashboard Scoped Filter project. This PR is for a new UI module: **filter scope selector modal**. It covers the UI/UX that display/select scope (tabs and charts) for a filter field. Update dashboard redux state and persist filter scope settings to backend, will be covered in another PR.

Superset dashboard stores layout data in a tree structure. This layout structure always contains a **root** component. A root component will have a grid component (for dashboard without tabs) or a root-level tabs components as its children. In the dashboard, every UI component (tabs, row, column, and chart, etc) is a **tree node**. For example:
![61979193-0da82280-afa8-11e9-9735-cd5c4435cc1e](https://user-images.githubusercontent.com/27990562/67975697-d6ae6200-fbd1-11e9-8931-4eac27c198d6.jpg)
We can think A is root component, leaves(K,L, F, G, M, I, J) are chart components, and the rest can be container (like row, column or tabs) components.

With the growing size of the dashboard, user may want to apply one dashboard filter to some of charts in dashboard, and one filter apply to another set of charts. A very common use case is user groups similar charts in different tabs, and make a filter only applicable for this group. Currently the way to set filter's scope is not very easy to use:
<img width="1140" alt="Screen_Shot_2019-10-16_at_9_49_54_AM" src="https://user-images.githubusercontent.com/27990562/66944436-00785e00-f002-11e9-81c1-5b16aa6ba6ae.png">

This PR introduced a new UI component: filter scope selector.

- **left panel**: All filter_box in dashboard and their filter fields. There is also a toggle link to switch between single-edit mode or multi-edit mode.
- **right panel**: show dashboard layout in the tree structure. But we hide row/column container, and group charts into its tab parent. If user have nested tabs containers, we will show all the nested tabs structure from chart to dashboard root component. We show a checkbox for each of chart and tab, user can click to select or deselect single or multiple charts. 

When user clicks the `Save` button at the bottom of the modal, we will **aggregate** the checked chart ids **into _scope_ object** and update dashboard filters state. The algorithm for this aggregation will be covered in another PR. 

#### Features:
1. Start filter scope selector:
![jwMsJ1jNea](https://user-images.githubusercontent.com/27990562/69028924-5e44f080-0988-11ea-8954-347fb5c89240.gif)

2. User can edit a single filter, or edit multiple filter fields all together. We also allow user toggle back to a single filter field in the multi-edit mode:
![PtIvVuSV6s](https://user-images.githubusercontent.com/27990562/69029436-c47e4300-0989-11ea-9368-b0bda0413b80.gif)

3. Search chart name:
![d2ZfBriTy2](https://user-images.githubusercontent.com/27990562/69029719-9baa7d80-098a-11ea-9f3f-e8dbda3f53b0.gif)

4. When new chart is added to dashboard, based on its layout position, it will get filters that are applicable to its tab(s):
![NvggG46SWb](https://user-images.githubusercontent.com/27990562/69032554-a7e60900-0991-11ea-8d9c-14fe0dbcd53b.gif)

5. When user move charts between tabs, this may cause chart to be applicable to different filter(s). We will show a warning message when chart is moved to a different filter scope.
![MVYXjUXxDr](https://user-images.githubusercontent.com/27990562/69031530-25f4e080-098f-11ea-9051-bfd7d77be1ae.gif)


### Implementation

Filter scope selector has following UI components:
- Container component: `src/dashboard/containers/FilterScope.jsx`
- modal trigger: `src/dashboard/components/filterscope/FilterScopeModal.jsx`
- core function UI component: `src/dashboard/components/filterscope/FilterScopeSelector.jsx`
- checkbox tree component in left pane: `FilterFieldTree.jsx`
recursive traversal renderer for each filter field tree node: `renderFilterFieldTreeNodes.jsx`
simple UI component for filter filed item: `FilterFieldItem.jsx`

- checkbox tree component in right pane:  `FilterScopeTree.jsx`
recursive traversal renderer for each filter scope tree node: `renderFilterScopeTreeNodes.jsx`

- util functions:

`getFilterFieldNodesTree`: build filter field tree from dashboard filters redux state.
`getFilterScopeNodesTree`: build filter scope tree from dashboard layout.

#### Tree and tree nodes
We used [a 3rd party component](https://github.com/jakezatecky/react-checkbox-tree) to render left panel filter field selector and right panel scope selector. It best matched our layout data structure, and offers very natural and convenient user experience like single selection, group selection, expand-on-click, etc. 

According to [this API](https://github.com/jakezatecky/react-checkbox-tree#properties), a`CheckboxTree` component needs following props:
```
// the original data to be rendered in `CheckboxTree`. The data should be in tree structure. Since js doesn't have build-in Tree data structure, we use array of node object.
// for filter scope tree, it is full list of the dashboard layout (only have tabs and chart components), 
// for filter fields tree, it is all dashboard filter box and its fields.
nodes: PropTypes.arrayOf(filterScopeSelectorTreeNodePropShape).isRequired,

// when user typed keywords in search box, right pane will only show matched charts and all their parent tabs. Only used by filter scope tree component.
nodesFiltered: PropTypes.arrayOf(filterScopeSelectorTreeNodePropShape).isRequired,

// An array of  node values. `CheckboxTree` component will show these node as checked.
checked: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  ).isRequired, 

// An array of node values. `CheckboxTree` component will show these node as expanded tree nodes. 
expanded: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  ).isRequired
```

For filter field tree, the tree node data is:
```
{
  value: filter key.
  label: customized column label or column name for the chosen filter field.
  children: array of children node
}
``` 
For filter scope tree, the tree node data is:
```
{
  value: chart id or tab component id
  label: chart name or tab title
  children: array of children node
}
``` 

#### Local state in the filter scope selector modal
For each filter filed, we construct a unique filter key which is composed of filter_box's chart id and filter's column name: _`chartId_columnName`_

`FilterScopeSelector` component uses following data to maintain the **current state in the modal:**

- `showSelector`: as long as dashboard has one or more filter, we show filter scope selector UI. otherwise, show message says no filter in the dashboard.
- `filterFieldNodes`: filter fields tree object.
- `filterScopeMap`: a map for filter scope selections. in each entry, key is filter key, value is filter scope tree object for a filter field.
- `checkedFilterFields`: array of all chosen filter key(s) in the left pane.

Based on the chosen filter keys from left pane, we show the filter scope tree in the right pane, which memorize the expanded tabs and selected charts. If all charts under the tab are selected, tab will show **checked** state in its checkbox. If some of the charts under the tab are selected, tab will show **half checked** state. And if none of the charts are selected, tab will show a **empty** state for its checkbox.


### TEST PLAN
Manual test.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [X] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@etr2460 @williaster @mistercrunch @kenchendesign

Commit:
filter scope selector modal

Commit:
filter scope selector modal

dpgaspar committed:
[docs] CTAS on PostgreSQL needs commit to apply (#8367)

* [docs] New, document need for PG to use autocommit for CTAS

dpgaspar committed:
[docs] CTAS on PostgreSQL needs commit to apply (#8367)

* [docs] New, document need for PG to use autocommit for CTAS

Commit:
[Config] Cache static resources (#8370)

Commit:
[Config] Cache static resources (#8370)

john-bodley committed:
[sqla] Enforcing ISO 8601 date/timestamp formats (#7702)

john-bodley committed:
[sqla] Enforcing ISO 8601 date/timestamp formats (#7702)

john-bodley committed:
[ci] Deprecate flake8 (#8409)

* [ci] Deprecate flake8

* Addressing @villebro's comments

john-bodley committed:
[ci] Deprecate flake8 (#8409)

* [ci] Deprecate flake8

* Addressing @villebro's comments

francishemingway committed:
fixed typo in installation instructions (#8413)

francishemingway committed:
fixed typo in installation instructions (#8413)

MarcusSorealheis committed:
[tests] Fix, removing deprecated function for future Python release compatibility. (#8353)

* removing deprecated function for future Python release compatibility.

* also removed assertNotEquals per @dpgaspar

MarcusSorealheis committed:
[tests] Fix, removing deprecated function for future Python release compatibility. (#8353)

* removing deprecated function for future Python release compatibility.

* also removed assertNotEquals per @dpgaspar

betodealmeida committed:
Add improved typed casting to BigQuery (#8331)

betodealmeida committed:
Add improved typed casting to BigQuery (#8331)

Commit:
convert ids to scope object

Commit:
convert ids to scope object

etr2460 commented on superset/assets/src/dashboard/util/getFilterScopeNodesTree.js:
```diff
@@ -0,0 +1,110 @@
+    let children = [];
+    if (currentNode.children && currentNode.children.length) {
+      currentNode.children.forEach(child => {
+        const cNode = traverse(components[child]);
```
what's a `cNode`? full variable names please

graceguo-supercat replied:
fixed.

etr2460 commented on superset/assets/src/dashboard/util/getFilterScopeParentNodes.js:
```diff
@@ -0,0 +1,39 @@
+    }
+  };
+
+  if (nodes && nodes.length) {
```
prefer `nodes.length > 0`

graceguo-supercat replied:
fixed.

etr2460 reviewed (changes requested):
first pass!

mistercrunch committed:
chore: disable another flaky cypress test (#8423)

Also fix isort issue on master

mistercrunch committed:
chore: disable another flaky cypress test (#8423)

Also fix isort issue on master

john-bodley committed:
[pydruid] Bumping the pydruid version (#8430)

john-bodley committed:
[pydruid] Bumping the pydruid version (#8430)

kristw commented on superset/assets/src/dashboard/components/filterscope/FilterScopeModal.jsx:
```diff
@@ -0,0 +1,62 @@
+
+    this.modal = null;
+    this.close = this.close.bind(this);
+    this.setModalRef = this.setModalRef.bind(this);
```
can use `React.createRef()` instead of manually setting `ref`

graceguo-supercat replied:
fixed.

kristw commented on superset/assets/src/dashboard/components/filterscope/FilterScopeTree.jsx:
```diff
@@ -0,0 +1,71 @@
+      expanded={expanded}
+      onCheck={onCheck}
+      onExpand={onExpand}
+      onClick={() => {}}
```
same with `icons`

graceguo-supercat replied:
> same with `icons`

what does this mean? icons used in this file are outside of the render function.

graceguo-supercat replied:
fixed.

kristw reviewed (commented):
I wonder if in the future we can break a PR like this into smaller chunks? It is difficult to review 1600 lines of code.
Perhaps small PRs for the individual UI components and util functions, before another PR for the integration.

williaster commented on superset/assets/src/dashboard/components/filterscope/FilterFieldTree.jsx:
```diff
@@ -0,0 +1,72 @@
+      onClick={onClick}
+      onCheck={onCheck}
+      onExpand={onExpand}
+      icons={{
```
can you move these outside the render function so they aren't re-created each render invocation?

graceguo-supercat replied:
yes. fixed.

williaster commented on superset/assets/src/dashboard/components/filterscope/FilterScopeModal.jsx:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  close() {
+    this.modal.close();
```
can you add a check that `this.modal` is defined?

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/src/dashboard/components/filterscope/FilterScopeTree.jsx:
```diff
@@ -0,0 +1,71 @@
+      onCheck={onCheck}
+      onExpand={onExpand}
+      onClick={() => {}}
+      icons={{
```
similarly can you pull these out into a `const` outside the render function?

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/src/dashboard/components/filterscope/renderFilterFieldTreeNodes.jsx:
```diff
@@ -0,0 +1,46 @@
+import { getFilterColorMap } from '../../util/dashboardFiltersColorMap';
+
+export default function renderFilterFieldTreeNodes({ nodes, activeKey }) {
+  if (nodes.length === 0) {
```
you could add a `if (!nodes) return []` instead?

graceguo-supercat replied:
fixed. add a `if (!nodes) return []`

williaster commented on superset/assets/src/dashboard/components/filterscope/renderFilterFieldTreeNodes.jsx:
```diff
@@ -0,0 +1,46 @@
+    ...node,
+    children: node.children.map(child => {
+      const { label, value } = child;
+      const colorCode = getFilterColorMap()[value];
```
what if `getFilterColorMap()` returns nothing?

williaster commented on superset/assets/src/dashboard/components/filterscope/renderFilterScopeTreeNodes.jsx:
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+
+export default function renderFilterScopeTreeNodes(nodes) {
+  if (nodes.length === 0) {
```
ditto `if (!nodes) ...`, or define a default `nodes = []`

graceguo-supercat replied:
fixed. added default.

williaster commented on superset/assets/src/dashboard/reducers/getInitialState.js:
```diff
@@ -199,6 +198,8 @@ export default function(bootstrapData) {
     }
   });
+  buildActiveFilters(dashboardFilters);
+  buildFilterColorMap(dashboardFilters);
```
yeah maybe add a comment for why if it's important? (e.g., if the layout `sliceName` update on line `198` is important)

graceguo-supercat replied:
yes. this is important change. if you compared with old version, these 2 lines:
```
  buildActiveFilters(dashboardFilters);
  buildFilterColorMap(dashboardFilters);
```
was inside `dashboard.slices.forEach` loop. In the loop we build the initial state for dashboardFilters. These 2 lines are used to get all currently active filter's values and pick a color badge for each filter field. I think it should be called only once (after all dashboardFilter entries are populated from bootstrap data).

williaster commented on superset/assets/src/dashboard/stylesheets/filter-scope-selector.less:
```diff
@@ -0,0 +1,245 @@
+
+      &.chart {
+        &::before {
+          content: 'Chart';
```
++

graceguo-supercat replied:
fixed.

williaster commented on superset/assets/src/dashboard/util/activeDashboardFilters.js:
```diff
@@ -17,14 +17,31 @@
  * under the License.
  */
 let activeFilters = {};
+let allFilterIds = [];
```
++

graceguo-supercat replied:
rename it to `allFilterBoxChartIds`.

williaster commented on superset/assets/src/dashboard/util/getCurrentScopeChartIds.js:
```diff
@@ -0,0 +1,62 @@
+
+import { CHART_TYPE } from '../util/componentTypes';
+
+export default function getCurrentScopeChartIds({
```
nit -- `getChartIdsInCurrentScope`

williaster replied:
agree that without types it'd be great to add a function description that says what this does for better readability.

graceguo-supercat replied:
The purpose of this function is to find associated chart Ids  given a _scope_ object. This function was refactored during adding filter scope logic into dashboard filters redux state. This function is actually will be re-used by more components. And filterImmuneSlices and filterImmuneSliceFields should be merged into initial dashboard filters state.

Let's re-visit this function in next PR.

williaster commented on superset/assets/src/dashboard/util/getDashboardFilterKey.js:
```diff
@@ -0,0 +1,27 @@
+  return `${chartId}_${column}`;
+}
+
+export function getDashboardFilterByKey(key) {
```
I don't understand why this function is called `getDashboardFilterByKey`, maybe `getChartIdAndColumnFromFilterId`?

graceguo-supercat replied:
renamed to `getChartIdAndColumnFromFilterKey`

williaster commented on superset/assets/src/dashboard/util/getDashboardFilterKey.js:
```diff
@@ -0,0 +1,27 @@
+export function getDashboardFilterByKey(key) {
+  const [chartId, ...parts] = key.split('_');
+  const columnName = parts.slice().join('_');
+  return [parseInt(chartId, 10), columnName];
```
it also seems safer to return an object with keys `{ chartId, columnName }` vs an array with arbitrary length and items.

graceguo-supercat replied:
yes, object with keys is a lot safer and more readable. fixed.

williaster commented on superset/assets/src/dashboard/util/getDashboardFilterKey.js:
```diff
@@ -0,0 +1,27 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+export function getDashboardFilterKey(chartId, column) {
```
can you change this to named arguments for better type safety + readability? (you use named arguments in many other places)

graceguo-supercat replied:
fixed as much as i found.

williaster commented on superset/assets/src/dashboard/components/filterscope/renderFilterFieldTreeNodes.jsx:
```diff
@@ -0,0 +1,46 @@
+import FilterFieldItem from './FilterFieldItem';
+import { getFilterColorMap } from '../../util/dashboardFiltersColorMap';
+
+export default function renderFilterFieldTreeNodes({ nodes, activeKey }) {
```
should these have default values or should you check if they are defined below?

graceguo-supercat replied:
added default values for nodes and activeKey

williaster commented on superset/assets/src/dashboard/util/getFilterScopeNodesTree.js:
```diff
@@ -0,0 +1,110 @@
+
+const FILTER_SCOPE_CONTAINER_TYPES = [TAB_TYPE, DASHBOARD_ROOT_TYPE];
+
+export default function getFilterScopeNodesTree({
```
can you please add a description for what this function does?

etr2460 replied:
bumping this comment!

graceguo-supercat replied:
please see the `Implementation` section. This tree structure is used by react-checkbox-tree framework.

williaster reviewed (commented):
left some comments, will do another pass later

mistercrunch committed:
docs: add a note to RELEASING about Slack Channel (#8419)

mistercrunch committed:
docs: add a note to RELEASING about Slack Channel (#8419)

mistercrunch committed:
docs: Add an 'Orientation' section to CONTRIBUTING.md (#8422)

mistercrunch committed:
docs: Add an 'Orientation' section to CONTRIBUTING.md (#8422)

Commit:
remove changes that should be another PR

Commit:
remove changes that should be another PR

villebro committed:
[fix] refine release instructions (#8381)

* Separate RC from VERSION

* Fix pypi_push.sh and refine instructions

* Add SUPERSET_ prefix to env variables

* Finalize release instructions

* Change one-off to First Time Only

* Add tagging of final version

* Convert some remaining env variables and add a check step to pypi deploy

villebro committed:
[fix] refine release instructions (#8381)

* Separate RC from VERSION

* Fix pypi_push.sh and refine instructions

* Add SUPERSET_ prefix to env variables

* Finalize release instructions

* Change one-off to First Time Only

* Add tagging of final version

* Convert some remaining env variables and add a check step to pypi deploy

john-bodley committed:
[fix] Updating parse_human_timedelta typing (#8436)

john-bodley committed:
[fix] Updating parse_human_timedelta typing (#8436)

austinpray committed:
spelling fix (#8442)

austinpray committed:
spelling fix (#8442)

christophlingg committed:
explain the need to enable async queries (#8444)

christophlingg committed:
explain the need to enable async queries (#8444)

christophlingg committed:
introduce a space in command line option (#8438)

see https://docs.celeryproject.org/en/latest/userguide/optimizing.html

christophlingg committed:
introduce a space in command line option (#8438)

see https://docs.celeryproject.org/en/latest/userguide/optimizing.html

Commit:
Split up tests/db_engine_test.py (#8449)

* Split up db_engine_specs_test.py into a number of targeted files

* Remove db_engine_specs_test.py

* isort

Commit:
Split up tests/db_engine_test.py (#8449)

* Split up db_engine_specs_test.py into a number of targeted files

* Remove db_engine_specs_test.py

* isort

dpgaspar committed:
[setup] Fix, download_url (#8439)

* [setup] Fix, download_url

dpgaspar committed:
[setup] Fix, download_url (#8439)

* [setup] Fix, download_url

betodealmeida committed:
Allow fetching all rows from results endpoint (#8389)

* Allow bypassing DISPLAY_MAX_ROW

* Add unit tests and docs

* Fix tests

* Fix mock

* Fix unit test

* Revert config change after test

* Change behavior

* Address comments

betodealmeida committed:
Allow fetching all rows from results endpoint (#8389)

* Allow bypassing DISPLAY_MAX_ROW

* Add unit tests and docs

* Fix tests

* Fix mock

* Fix unit test

* Revert config change after test

* Change behavior

* Address comments

Commit:
add single-field-edit in multi-edit mode switch

Commit:
add single-field-edit in multi-edit mode switch

alexandrebrilhante committed:
removed repeated word (#8458)

alexandrebrilhante committed:
removed repeated word (#8458)

dpgaspar committed:
[docs] Crafting a new release (#8433)

* [docs] Crafting a new release

dpgaspar committed:
[docs] Crafting a new release (#8433)

* [docs] Crafting a new release

Commit:
Add UI-only database configuration method for extended authorization scenarios (#8448)

* Add encrypted_extra to dbs

* WIP - UI-based BigQuery connection configuration

* Fix 500 bubbling to the surface when adding a database connection

* Add check for valid json

* black formatting:

* isort

* Incorporate PR feedback. Thanks all!

* black

* Typo fix in CONTRIBUTING.md

Commit:
Add UI-only database configuration method for extended authorization scenarios (#8448)

* Add encrypted_extra to dbs

* WIP - UI-based BigQuery connection configuration

* Fix 500 bubbling to the surface when adding a database connection

* Add check for valid json

* black formatting:

* isort

* Incorporate PR feedback. Thanks all!

* black

* Typo fix in CONTRIBUTING.md

dpgaspar committed:
[db engine] Add support for Elasticsearch (#8441)

* [db engine] Add support for Elasticsearch

dpgaspar committed:
[db engine] Add support for Elasticsearch (#8441)

* [db engine] Add support for Elasticsearch

john-bodley committed:
[SIP-15] Making client time use UTC as the local time (#8450)

* [fix] Making client time UTC

* Update UPDATING.md

john-bodley committed:
[SIP-15] Making client time use UTC as the local time (#8450)

* [fix] Making client time UTC

* Update UPDATING.md

john-bodley committed:
[SIP-15] Adding initial framework (#8398)

* [sip-15] Adding initial framework

* [toast] Addressing etr2460's comments

* [fix] Addressing etr2460's comments

john-bodley committed:
[SIP-15] Adding initial framework (#8398)

* [sip-15] Adding initial framework

* [toast] Addressing etr2460's comments

* [fix] Addressing etr2460's comments

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterFieldTree.jsx:
```diff
@@ -0,0 +1,72 @@
+
+const propTypes = {
+  activeKey: PropTypes.string.isRequired,
+  nodes: PropTypes.arrayOf(PropTypes.object).isRequired,
```
Actually I tried to add a shape for tree node type. it should be something like this:
```
export const filterScopeNodePropShape = PropTypes.shape({
  children: PropTypes.arrayOf(filterScopeNodePropShape),
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
});
```
The `children` attribute is an array of its own type. JS Lint complaints `no-use-before-define`. if anyone has right way to fix this I will be really appreciated.

kristw replied:
```js
// Declare PropTypes for recursive data structures
// https://github.com/facebook/react/issues/5676
/* eslint-disable-next-line  no-undef */
const lazyFunction = f => () => f().apply(this, arguments);

const leafType = PropTypes.shape({
  name: PropTypes.string,
  value: PropTypes.number.isRequired,
});

const parentShape = {
  name: PropTypes.string,
  children: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.shape(lazyFunction(() => parentShape)), leafType]),
  ),
};

const nodeType = PropTypes.oneOfType([PropTypes.shape(parentShape), leafType]);
```

graceguo-supercat replied:
fixed!!

Commit:
fix code review comments (round 1)

Commit:
fix code review comments (round 1)

graceguo-supercat commented on superset/assets/spec/javascripts/dashboard/reducers/dashboardFilters_spec.js:
```diff
@@ -35,7 +35,8 @@ import {
 
-describe('dashboardFilters reducer', () => {
+// disable broken unit tests by now, will fix it in another PR
+xdescribe('dashboardFilters reducer', () => {
```
fixed. this is dashboardFilters reducer unit test. I have a lot more changes in next PR, which is  to merge filter scope logic with dashboardFilters redux state. I keep minimum change for redux change in this PR to avoid merge conflicts.

Commit:
Provide documentation for using a Service Account to connect to BigQuery (#8462)

* Provide documentation for using a Service Account to connect to BigQuery

* Alter line wrapping for shorter lines

* Whitespace commit to trigger another build (flake)

* Another meaningless whitespace change to trigger another build

Commit:
Provide documentation for using a Service Account to connect to BigQuery (#8462)

* Provide documentation for using a Service Account to connect to BigQuery

* Alter line wrapping for shorter lines

* Whitespace commit to trigger another build (flake)

* Another meaningless whitespace change to trigger another build

Commit:
[Datasource Editor] A few small UI changes in modal to prevent accidental edits (#8471)

Commit:
[Datasource Editor] A few small UI changes in modal to prevent accidental edits (#8471)

graceguo-supercat commented on superset/assets/src/components/CheckboxChecked.jsx:
```diff
@@ -0,0 +1,28 @@
+      <path d="M7 14L2 9L3.41 7.59L7 11.17L14.59 3.58L16 5L7 14Z" fill="white" />
+    </svg>
+  );
+}
```
fixed. i created single file holding 3 svg icons, and create a const for icons props.

graceguo-supercat commented on superset/assets/src/components/CheckboxHalfchecked.jsx:
```diff
@@ -0,0 +1,28 @@
+ */
+import React from 'react';
+
+export default function CheckboxHalfchecked() {
```
fixed. CheckboxHalfChecked

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+  layout: PropTypes.object.isRequired,
+  filterImmuneSlices: PropTypes.arrayOf(PropTypes.number).isRequired,
+  filterImmuneSliceFields: PropTypes.object.isRequired,
+
```
fixed.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+  onSave() {
+    const { filterScopeMap } = this.state;
+
+    console.log(
```
i plan to remove console log when integrate `FilterScopeSelector` component to dashboardFilters reducers.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+      this.defaultFilterKey = Object.keys(filterFieldNodes).length
+        ? filterFieldNodes[0].children[0].value
+        : '';
+      const checkedFilterFields = [this.defaultFilterKey];
```
filterFieldNodes should be array type. Object.keys(filterFieldNodes) check is wrong. fixed.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+
+      this.defaultFilterKey = Object.keys(filterFieldNodes).length
+        ? filterFieldNodes[0].children[0].value
+        : '';
```
i prefer to initialize primitive type with primitive value.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+                  // filtered nodes in display if searchText is not empty
+                  nodesFiltered: nodes.slice(),
+                  checked: getCurrentScopeChartIds({
+                    scopeComponentIds: ['ROOT_ID'], // dashboardFilters[chartId].scopes[columnName],
```
this `scopeComponentIds` should read from redux state `dashboardFilters[chartId].scopes[columnName]`, but i don't have redux state change in this PR yet. use `['ROOT_ID']` as mock data.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+                  // unfiltered nodes
+                  nodes,
+                  // filtered nodes in display if searchText is not empty
+                  nodesFiltered: nodes.slice(),
```
fixed.

etr2460 replied:
Looks like this fix got lost in a rebase or commit somewhere

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+    this.onSave = this.onSave.bind(this);
+  }
+
+  onCheckFilterScope(checked) {
```
yes. I added default values for most of functions.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+    if (isSingleEditMode) {
+      const updatedEntry = {
+        ...filterScopeMap[activeKey],
+        checked: checked.map(c => JSON.parse(c)),
```
https://github.com/jakezatecky/react-checkbox-tree#render-component

The original value's type is number, i still need to keep its original type when i update state. But onCheck function return a list of serialized values, number became string.

john-bodley committed:
Update UPDATING.md (#8465)

john-bodley committed:
Update UPDATING.md (#8465)

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+    }));
+  }
+
+  onClickFilterField(filterField) {
```
cool. fixed.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+      searchText,
+    } = this.state;
+    return (
+      <React.Fragment>
```
No. in all Superset codebase we only use <React.Fragment>. I prefer keep code style consistent.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+          className={cx('filter-text scope-search', {
+            'multi-edit-mode': !isSingleEditMode,
+          })}
+          placeholder="Search..."
```
fixed.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+            <h4>{t('Configure filter scopes')}</h4>
+          </div>
+
+          {!showSelector && <div>There is no filter in this dashboard</div>}
```
fixed.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+            <h4>{t('Configure filter scopes')}</h4>
+          </div>
+
+          {!showSelector && <div>There is no filter in this dashboard</div>}
```
fixed.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/renderFilterScopeTreeNodes.jsx:
```diff
@@ -0,0 +1,52 @@
+    return [];
+  }
+
+  function traverse(currentNode) {
```
fixed.

john-bodley committed:
[sql] Fixing datetime SQL literal (#8464)

john-bodley committed:
[sql] Fixing datetime SQL literal (#8464)

graceguo-supercat commented on superset/assets/src/dashboard/containers/FilterScope.jsx:
```diff
@@ -0,0 +1,48 @@
+    filterImmuneSliceFields:
+      dashboardInfo.metadata.filterImmuneSliceFields || {},
+    layout: dashboardLayout.present,
+    // closeModal: ownProps.onCloseModal,
```
seems not necessary. so I removed the whole comment.

graceguo-supercat commented on superset/assets/src/dashboard/reducers/dashboardFilters.js:
```diff
@@ -32,11 +31,13 @@ import { buildActiveFilters } from '../util/activeDashboardFilters';
 export const dashboardFilter = {
   chartId: 0,
   componentId: '',
+  filterName: '',
```
i prefer to initialize primitive types with primitive value.

etr2460 replied:
Personally, I think that empty string has a very different connotation in code than null. Defaulting to null means that a value doesn't exist or was never set. Defaulting to empty string means that this value does exist and was set to `''`.

@williaster or @kristw, do you have any opinions on this? What have we been doing in other parts of Superset/superset-ui?

williaster replied:
In TS it is very common to set to `null` to signal that a value is unset, vs set to the empty string. I definitely understand the argument for wanting to initialize primitive types to the matching primitive values.

@kristw wdyt?

graceguo-supercat commented on superset/assets/src/dashboard/util/getFilterFieldNodesTree.js:
```diff
@@ -0,0 +1,43 @@
+  dashboardFilters = {},
+  isSingleEditMode = true,
+}) {
+  if (Object.keys(dashboardFilters).length === 0) {
```
cool. fixed.

mistercrunch committed:
docs: more release instructions (#8432)

mistercrunch committed:
docs: more release instructions (#8432)

khtruong committed:
feat: add expand_data parameter (#8472)

* feat: add expand_data parameter

* fix: reformat files

khtruong committed:
feat: add expand_data parameter (#8472)

* feat: add expand_data parameter

* fix: reformat files

Commit:
Use `config[]` not `config.get()` (#8454)

* Typo fix in CONTRIBUTING.md

* Alter references to config.get('FOO') to use preferred config['FOO']

* Set missing configuration constants in superset/config.py

* Misc. CI fixes

* Add type annotation for FEATURE_FLATGS

Commit:
Use `config[]` not `config.get()` (#8454)

* Typo fix in CONTRIBUTING.md

* Alter references to config.get('FOO') to use preferred config['FOO']

* Set missing configuration constants in superset/config.py

* Misc. CI fixes

* Add type annotation for FEATURE_FLATGS

graceguo-supercat commented on superset/assets/spec/javascripts/dashboard/reducers/dashboardFilters_spec.js:
```diff
@@ -54,7 +55,7 @@ describe('dashboardFilters reducer', () => {
         componentId: component.id,
         directPathToFilter,
-        scope: DASHBOARD_ROOT_ID,
+        scope: 'ROOT_ID',
```
fixed the unit test.

graceguo-supercat replied:
fixed the whole unit test.

Commit:
Re-enable DRUID_IS_ACTIVE flag (#8482)

* Re-enable DRUID_IS_ACTIVE flag

* Make CI happy

Commit:
Re-enable DRUID_IS_ACTIVE flag (#8482)

* Re-enable DRUID_IS_ACTIVE flag

* Make CI happy

MarcusSorealheis committed:
fixing typo. (#8487)

MarcusSorealheis committed:
fixing typo. (#8487)

christophlingg committed:
Enhance docs for email reporting (#8486)

* extend documentation about setting up email reporting

* mention EMAIL_REPORTS_USER

christophlingg committed:
Enhance docs for email reporting (#8486)

* extend documentation about setting up email reporting

* mention EMAIL_REPORTS_USER

graceguo-supercat commented on superset/assets/src/dashboard/stylesheets/dashboard.less:
```diff
@@ -165,19 +165,26 @@ body {
-      }
+  .dashboard-modal.delete {
+    .btn.btn-primary {
+      background: @pink !important;
```
no need. fixed.

graceguo-supercat commented on superset/assets/src/dashboard/stylesheets/filter-scope-selector.less:
```diff
@@ -0,0 +1,245 @@
+@import "../../../stylesheets/less/cosmo/variables.less";
+
+.filter-scope-container {
+  font-family: @font-family-sans-serif;
```
not necessary. removed.

graceguo-supercat commented on superset/assets/src/dashboard/stylesheets/filter-scope-selector.less:
```diff
@@ -0,0 +1,245 @@
+      display: block;
+    }
+
+    // disable style from react-checkbox-tress.css
```
fixed.

john-bodley committed:
[sip-15] Fix time range endpoints encoding (#8481)

john-bodley committed:
[sip-15] Fix time range endpoints encoding (#8481)

john-bodley committed:
[sip-15] Adding database level python-date-format (#8474)

john-bodley committed:
[sip-15] Adding database level python-date-format (#8474)

graceguo-supercat commented on superset/assets/src/dashboard/util/getDashboardFilterKey.js:
```diff
@@ -0,0 +1,27 @@
+
+export function getDashboardFilterByKey(key) {
+  const [chartId, ...parts] = key.split('_');
+  const columnName = parts.slice().join('_');
```
i am not sure which separator is kind of easy to read, and can guaranteed not used as content.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+        };
+      });
+
+      return;
```
fixed

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,481 @@
+
+  filterNodes(filtered, node) {
+    const { searchText } = this.state;
+    const children = (node.children || []).reduce(this.filterNodes, []);
```
Theoretically all recursion can be written in loop. But here recursion is really easier than regular loop. Here we use array type to represent a **Tree** data structure, because we really care about the hierarchy of the charts and its parents. The filter function is to find leaf node that matches search keyword, as well as all its parents tab up to root component. Recursive traverse is the best way to do this job.

conglei committed:
feat(TableVis): Switch to the new table plugin (#8446)

* use new table comp

* remove commented code

* update lock file

* fix lint

* bump version

* update lock file

* fix lock issue

conglei committed:
feat(TableVis): Switch to the new table plugin (#8446)

* use new table comp

* remove commented code

* update lock file

* fix lint

* bump version

* update lock file

* fix lock issue

Commit:
[datasource editor] Only one click target for edit action (#8495)

Commit:
[datasource editor] Only one click target for edit action (#8495)

john-bodley committed:
[SIP-15] Updating ISO 8601 regex match to make time optional (#8494)

john-bodley committed:
[SIP-15] Updating ISO 8601 regex match to make time optional (#8494)

francishemingway committed:
Change Fedora installation instructions + some small formatting changes (#8496)

francishemingway committed:
Change Fedora installation instructions + some small formatting changes (#8496)

yahwang committed:
feat: add korea(South) info on CountryMap viz and bump legacy-plugin-chart-country-map to v0.11.4 (#8478)

yahwang committed:
feat: add korea(South) info on CountryMap viz and bump legacy-plugin-chart-country-map to v0.11.4 (#8478)

paulvic committed:
Add `Duplicate Tab` option (#8485)

* Add `Duplicate Tab` option

Adds an option to duplicate the current tab's setting/content into a new tab. Useful if you're iterating on a query.

* Add test

paulvic committed:
Add `Duplicate Tab` option (#8485)

* Add `Duplicate Tab` option

Adds an option to duplicate the current tab's setting/content into a new tab. Useful if you're iterating on a query.

* Add test

etr2460 commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,518 @@
+                selectedChartId: filterId,
+              });
+              const checked = getCurrentScopeChartIds({
+                scopeComponentIds: ['ROOT_ID'], // dashboardFilters[chartId].scopes[columnName],
```
should this always be `'ROOT_ID'`? or is the commented part here supposed to be uncommented?

graceguo-supercat replied:
by default, or all new filter field, their scope should be all dashboard charts. But after user change/set new scope for a filter field, the new scope object will be stored in redux state (and saved to dashboard metadata in backend).
When user open filter scope selector, the saved filter scope will be pre-populated from redux store. This line is mock data for this component. It will be changed in the next PR.

etr2460 commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,518 @@
+      this.state = {
+        showSelector: true,
+        activeKey: this.defaultFilterKey,
+        searchText: '',
```
can we default this to `null` instead of empty string? That'll make some of the other checks along the way more explicit (e.g. `searchText !== null` instead of `!searchText`)

graceguo-supercat replied:
1. If i initialize the string type a string value, i will feel confident to do string operation on it, like append, truncate, split, and don't need to check and handle null case.
2. isn't `!searchText` cooler than `searchText !== null ` 😂
3. if there has no proven evidence about performance difference, why do we bother to change it?

williaster replied:
I think 2) and 3) aren't really significant concerns here.

using `null` is quite common esp in TypeScript. I agree that this allows for checking explicitly if there is a value or not, since there is difference in meaning between setting something to `null` vs `empty string`

graceguo-supercat replied:
fixed. use null as initial value

etr2460 replied:
i don't see the change in this PR. Did you forget to push?

etr2460 commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,518 @@
+          </div>
+
+          {!showSelector ? (
+            <div>{t('There are no filter in this dashboard')}</div>
```
grammar nit: `There are no filters in this dashboard`

graceguo-supercat replied:
fixed.

etr2460 commented on superset/assets/src/dashboard/stylesheets/filter-scope-selector.less:
```diff
@@ -0,0 +1,253 @@
+}
+
+.filters-scope-selector {
+  margin: 10px -24px 20px -24px;
```
this can be `margin: 10px -24px 20px;`

graceguo-supercat replied:
fixed.

etr2460 commented on superset/assets/src/dashboard/stylesheets/filter-scope-selector.less:
```diff
@@ -0,0 +1,245 @@
+    .rct-option .rct-icon {
+      &.rct-icon-expand-all {
+        &::before {
+          content: 'Expand all';
```
This still needs to be moved into a t string

etr2460 commented on superset/assets/src/dashboard/stylesheets/filter-scope-selector.less:
```diff
@@ -0,0 +1,245 @@
+
+      &.rct-icon-collapse-all {
+        &::before {
+          content: 'Collapse all';
```
This still needs to be moved into a t string

etr2460 commented on superset/assets/src/dashboard/util/activeDashboardFilters.js:
```diff
@@ -17,14 +17,31 @@
 // non-empty filters from dashboardFilters,
 // this function does not take into account: filter immune or filter scope settings
 export function buildActiveFilters(allDashboardFilters = {}) {
+  allFilterBoxChartIds = Object.values(allDashboardFilters).map(
```
Does this mean that we need to call `buildActiveFilters` before `getAllFilterBoxChartIds` and `isFilterBox` works? This sort of mutability is really error prone, is there a better way to do this?

graceguo-supercat replied:
it's called in getInitState and reducers function, whenever dashboardFilters state is changed. This closure is trying to simulate a singleton design pattern.

etr2460 commented on superset/assets/src/dashboard/util/getFilterScopeNodesTree.js:
```diff
@@ -0,0 +1,110 @@
+    if (FILTER_SCOPE_CONTAINER_TYPES.includes(type)) {
+      let label = '';
+      if (type === DASHBOARD_ROOT_TYPE) {
+        label = 'All dashboard';
```
should this be a t string? also, I think it should probably be `All dashboards`, but i can't tell the context here so not positive

graceguo-supercat replied:
fixed. Right now we have `Select/deselect all filters` and `Select/deselect all charts`. User can select or deselect all leaf nodes by clicking on the root.

etr2460 commented on superset/assets/src/dashboard/util/getFilterScopeNodesTree.js:
```diff
@@ -0,0 +1,110 @@
+  checkedFilterFields = [],
+  selectedChartId,
+}) {
+  function traverse(currentNode) {
```
i think `traverse` should be defined outside of this function so that it doesn't get recreated on every call

graceguo-supercat replied:
fixed.

etr2460 commented on superset/assets/src/dashboard/util/getFilterScopeParentNodes.js:
```diff
@@ -0,0 +1,39 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+export default function getFilterScopeParentNodes(nodes, depthLimit = 0) {
```
if depthLimit specifies a cutoff depth, then `-1` should probably indicate no limit, since 0 would be a valid depthLimit (returning just the original node)

graceguo-supercat replied:
fixed. ❤️

etr2460 reviewed (commented):
second pass!

MarcusSorealheis committed:
build: bump dompurify version because of nasty xss bypass. (#8498)

MarcusSorealheis committed:
build: bump dompurify version because of nasty xss bypass. (#8498)

serenajiang committed:
[typing] superset/models/core.py (#8284)

serenajiang committed:
[typing] superset/models/core.py (#8284)

betodealmeida committed:
Allow user to customize query cost estimate (#8470)

* Allow user to customize query estimate

* Add docs; run black

* Update docs with types

betodealmeida committed:
Allow user to customize query cost estimate (#8470)

* Allow user to customize query estimate

* Add docs; run black

* Update docs with types

john-bodley committed:
[SIP-15] Adding grace period (#8490)

john-bodley committed:
[SIP-15] Adding grace period (#8490)

liuyanpunk committed:
Updating pip with default timeout (#8492)

* Updating pip with default timeout

Most of the time, the pip default site is realiable but we do see timeouts pops out when network is busy or temporary unavaliable .

The default value was 15 and that's way too short. So adding --default-time and set it to 1000 should be long enough to mitigate

* Update Dockerfile

liuyanpunk committed:
Updating pip with default timeout (#8492)

* Updating pip with default timeout

Most of the time, the pip default site is realiable but we do see timeouts pops out when network is busy or temporary unavaliable .

The default value was 15 and that's way too short. So adding --default-time and set it to 1000 should be long enough to mitigate

* Update Dockerfile

Commit:
Fix column type on dbs.encrypted_extra, add instructions for testing migration downgrades (#8493)

* Fix column type on dbs.encrypted_extra

* Add instructions for testing migration downgrades

* Account for non-Postgres DBs in migration

* Use batch_alter_table to make SQLite happy

* Another CI-appeasing tweak

Commit:
Fix column type on dbs.encrypted_extra, add instructions for testing migration downgrades (#8493)

* Fix column type on dbs.encrypted_extra

* Add instructions for testing migration downgrades

* Account for non-Postgres DBs in migration

* Use batch_alter_table to make SQLite happy

* Another CI-appeasing tweak

paulvic committed:
Update query tab's title when saving a new query (#8484)

* Update query tab's title when saving a new query

Just like it does when you opt to update an existing query

* Remove trailing whitespace

paulvic committed:
Update query tab's title when saving a new query (#8484)

* Update query tab's title when saving a new query

Just like it does when you opt to update an existing query

* Remove trailing whitespace

dpgaspar committed:
[docs] Small fixes on the release docs (#8500)

* [docs] Small fixes on the release docs

* [docs] Fix, remove deprecated app parameter from fab create-admin

dpgaspar committed:
[docs] Small fixes on the release docs (#8500)

* [docs] Small fixes on the release docs

* [docs] Fix, remove deprecated app parameter from fab create-admin

john-bodley committed:
[SIP-15] Making sure the toast link opens in a new tab (#8503)

* [SIP-15] Ensuring toast opens in a new tab

* Update config.py

john-bodley committed:
[SIP-15] Making sure the toast link opens in a new tab (#8503)

* [SIP-15] Ensuring toast opens in a new tab

* Update config.py

betodealmeida committed:
Remove tabs (#8508)

betodealmeida committed:
Remove tabs (#8508)

Commit:
refactory after design review

Commit:
refactory after design review

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeModal.jsx:
```diff
@@ -0,0 +1,59 @@
+    this.close = this.close.bind(this);
+  }
+
+  close() {
```
fixed.

Commit:
Make Report a bug link open in a new tab (#8509)

Commit:
Make Report a bug link open in a new tab (#8509)

graceguo-supercat commented:
there are some UI updates based on designers review:
- merge single-edit mode + multi-edit mode into one view.
- update a chart icon
- add root element for filter field pane.

![Screen_Shot_2019-11-05_at_10_55_46_PM](https://user-images.githubusercontent.com/27990562/68275274-091bed00-0020-11ea-9f90-b9fecfacc17f.jpg)

conglei committed:
revert tablevis (#8515)

conglei committed:
revert tablevis (#8515)

john-bodley committed:
[sip-15] Fixing time range endpoints from dashboards (#8513)

john-bodley committed:
[sip-15] Fixing time range endpoints from dashboards (#8513)

Zanoellia committed:
Default page: remove title attribute to fall back on appbuilder.app_name (#8427)

Zanoellia committed:
Default page: remove title attribute to fall back on appbuilder.app_name (#8427)

villebro committed:
Replace hardcoded default colors with defaults in setupColors.js (#8510)

villebro committed:
Replace hardcoded default colors with defaults in setupColors.js (#8510)

villebro committed:
[fix] Improve csv upload functionality (#8457)

* [fix] csv upload when table metadata present

* Remove table from hive spec

* Move upload before table metadata creation

* Refine upload logic, dd unit tests and fix translations

* Use ALLOWED_EXTENSIONS from config

* Address review comments

* Fix error message grammar

* Add return type to hive csv upload and replace first with one_or_none

villebro committed:
[fix] Improve csv upload functionality (#8457)

* [fix] csv upload when table metadata present

* Remove table from hive spec

* Move upload before table metadata creation

* Refine upload logic, dd unit tests and fix translations

* Use ALLOWED_EXTENSIONS from config

* Address review comments

* Fix error message grammar

* Add return type to hive csv upload and replace first with one_or_none

etr2460 commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,518 @@
+      const checkedFilterFields = [];
+      const activeFilterField = this.defaultFilterKey;
+      // expand defaultFilterKey in filter field tree
+      const expandedFilterIds = [ALL_FILTERS].concat(chartId);
```
is chartId not included in `ALL_FILTERS` seems like it wouldn't be all the filters then

graceguo-supercat replied:
`ALL_FILTERS` is a node value. just like dashboard's root's value is `ROOT_ID`. 
`expandedFilterIds` expects an array of node values that should be displayed as expanded tree branch. I think `ALL_FILTERS` seems misleading here, so i rename it to be `ALL_FILTERS_ROOT`.

etr2460 commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -0,0 +1,518 @@
+      ),
+    );
+
+    // save do not close modal
```
grammar nit: `save does not close modal`

graceguo-supercat replied:
fixed.

etr2460 commented on superset/assets/src/dashboard/util/getFilterScopeNodesTree.js:
```diff
@@ -0,0 +1,122 @@
+  currentNode = {},
+  components = {},
+  filterFields = [],
+  selectedChartId = 0,
```
do we need default params here? it looks like they're all always defined

graceguo-supercat replied:
yeah, fixed. `selectedChartId` prop could be `number` or `null`.

etr2460 commented on superset/assets/src/dashboard/util/getSelectedChartIdForFilterScopeTree.js:
```diff
@@ -0,0 +1,53 @@
+          getChartIdAndColumnFromFilterKey(filterKey).chartId !== chartId,
+      )
+    ) {
+      return 0;
```
why do we just return 0 here? It looks like we're supposed to be returning a chartId.

If we're treating 0 as representing the absence of a chartId, you probably can guess that i'd prefer to return `null`. Regardless though, if you want to return 0 here, we probably should have a const named `DEFAULT_CHART_ID` and set it to 0. Then you could use that throughout the code base, including anywhere you set a default chart id in function arguments.

All that said, i really would prefer returning null here.

graceguo-supercat replied:
fixed.

etr2460 reviewed (commented):
just a few more comments, it also looks like there are some older comments that are still pending

john-bodley committed:
[docs] Adding Slice parameter documentation (#8531)

* [CONTRIBUTING.md] Adding Slice parameter documentation

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

john-bodley committed:
[docs] Adding Slice parameter documentation (#8531)

* [CONTRIBUTING.md] Adding Slice parameter documentation

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

francishemingway committed:
[docs] User tutorial (#8459)

* added user tutorial and rearranged docs hierarchy

* added license header, renamed admin tutorial file

* fixed image url issue

* Fix spelling

* Tweaks to text based on feedback

* guilabel and menu test

* added guilabel and upload csv section

* tidy up of rst formatting

francishemingway committed:
[docs] User tutorial (#8459)

* added user tutorial and rearranged docs hierarchy

* added license header, renamed admin tutorial file

* fixed image url issue

* Fix spelling

* Tweaks to text based on feedback

* guilabel and menu test

* added guilabel and upload csv section

* tidy up of rst formatting

dpgaspar committed:
[cli] Fix, import datasources exported by UI (#8477)

* [cli] Fix, import datasources exported by UI

dpgaspar committed:
[cli] Fix, import datasources exported by UI (#8477)

* [cli] Fix, import datasources exported by UI

john-bodley committed:
[docs] Updating slice parameter documentation. (#8535)

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

john-bodley committed:
[docs] Updating slice parameter documentation. (#8535)

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

Commit:
fix a few props initial value

Commit:
fix a few props initial value

etr2460 reviewed (commented):
i don't see some of the changes you commented as fixed, did they get pushed to this branch?

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterFieldTree.jsx:
```diff
@@ -0,0 +1,94 @@
+  activeKey: PropTypes.oneOfType([null, PropTypes.string]),
+  nodes: PropTypes.arrayOf(filterScopeSelectorTreeNodePropShape).isRequired,
+  checked: PropTypes.arrayOf(
+    PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
```
When i set initial value as null for an input control, React framework game me this suggestion:
<img width="1113" alt="Screen Shot 2019-11-10 at 8 51 35 PM" src="https://user-images.githubusercontent.com/27990562/68633453-c9815500-04a6-11ea-82f4-9e618144744b.png">

graceguo-supercat replied:
checked props expects array of `values` of the nodes in the tree, so that checkbox tree display the checkbox as checked, unchecked or half checked. Why it should be boolean?

graceguo-supercat commented:
When i set initial value as null for an input control, React framework game me this suggestion:
<img width="1113" alt="Screen Shot 2019-11-10 at 8 51 35 PM" src="https://user-images.githubusercontent.com/27990562/68633453-c9815500-04a6-11ea-82f4-9e618144744b.png">

nytai committed:
fab 2.2.0 (#8517)

nytai committed:
fab 2.2.0 (#8517)

paulvic committed:
Allow customization of documentation icon and text (#8507)

* Allow customization of documentation icon and text

* Set icon width to 100%

* Use double quotes for strings

paulvic committed:
Allow customization of documentation icon and text (#8507)

* Allow customization of documentation icon and text

* Set icon width to 100%

* Use double quotes for strings

axuew committed:
Add support for database engine SAP Hana (#8411)

* Add support for database engine SAP Hana

* Support hana services

Increase time, minute, and second

* Fix hana return string

* Fix formatting errors

axuew committed:
Add support for database engine SAP Hana (#8411)

* Add support for database engine SAP Hana

* Support hana services

Increase time, minute, and second

* Fix hana return string

* Fix formatting errors

dpgaspar committed:
[docs] 0.35.0 CHANGELOG (#8547)

* [docs] Update CHANGELOG with 0.35.0

* [docs] Fix, github md problem

dpgaspar committed:
[docs] 0.35.0 CHANGELOG (#8547)

* [docs] Update CHANGELOG with 0.35.0

* [docs] Fix, github md problem

Commit:
Upgrade superset-ui/time-format (#8541)

Commit:
Upgrade superset-ui/time-format (#8541)

rusackas committed:
Clickable checkbox labels (#8534)

* passing onClick prop to header with the existing onChange method.

* basic test checking that label click fires the onChange method.

* cleaning up stuff caught by linting.

rusackas committed:
Clickable checkbox labels (#8534)

* passing onClick prop to header with the existing onChange method.

* basic test checking that label click fires the onChange method.

* cleaning up stuff caught by linting.

betodealmeida committed:
Encode feature flags to JSON pessimistically (#8529)

* Encode feature flags to JSON pessimistically

* Add unit test

* Remove old imports

betodealmeida committed:
Encode feature flags to JSON pessimistically (#8529)

* Encode feature flags to JSON pessimistically

* Add unit test

* Remove old imports

john-bodley committed:
[fix] Re-cleanup legacy filters (#8523)

john-bodley committed:
[fix] Re-cleanup legacy filters (#8523)

villebro committed:
Add convert_dttm method to SnowflakeEngineSpec (#8551)

willbarrett commented:
Hey @graceguo-supercat - this PR is getting impossible to review due to the size of the delta. Should we close this one and open something fresh that's rebased off of latest master?

graceguo-supercat commented:
Please comment in #8553.

graceguo-supercat commented:
@willbarrett @etr2460 review it for a long time, he prefer to use this old PR.

Commit:
Merge remote-tracking branch 'origin/scope-selector-modal' into scope-selector-modal
